operator/controllers/operation_controller.go

```diff
@@ -20,34 +20,338 @@ import (
 	"context"
 
 	"github.com/go-logr/logr"
+	"github.com/pkg/errors"
+	apierrors "k8s.io/apimachinery/pkg/api/errors"
+	"k8s.io/apimachinery/pkg/util/uuid"
+	"k8s.io/client-go/tools/record"
+	operatorv1 "k8s.io/kubeadm/operator/api/v1alpha1"
+	operatorerrors "k8s.io/kubeadm/operator/errors"
+	"k8s.io/kubeadm/operator/operations"
+	capierrors "sigs.k8s.io/cluster-api/errors"
+	"sigs.k8s.io/cluster-api/util/patch"
 	ctrl "sigs.k8s.io/controller-runtime"
 	"sigs.k8s.io/controller-runtime/pkg/client"
-
-	operatorv1alpha1 "k8s.io/kubeadm/operator/api/v1alpha1"
 )
 
 // OperationReconciler reconciles a Operation object
 type OperationReconciler struct {
 	client.Client
-	Log logr.Logger
+	ManagerContainerName string
+	ManagerNamespace     string
+	AgentImage           string
+	MetricsRBAC          bool
+	Log                  logr.Logger
+	recorder             record.EventRecorder
 }
 
+// +kubebuilder:rbac:groups=core,resources=events,verbs=get;list;watch;create;patch
+// +kubebuilder:rbac:groups=core,resources=pods,verbs=get;list;watch;
+// +kubebuilder:rbac:groups=apps,resources=daemonsets,verbs=get;list;watch;create;update;patch;delete
 // +kubebuilder:rbac:groups=operator.kubeadm.x-k8s.io,resources=operations,verbs=get;list;watch;create;update;patch;delete
 // +kubebuilder:rbac:groups=operator.kubeadm.x-k8s.io,resources=operations/status,verbs=get;update;patch
 
+// SetupWithManager configures the controller for calling the reconciler
+func (r *OperationReconciler) SetupWithManager(mgr ctrl.Manager) error {
+	err := ctrl.NewControllerManagedBy(mgr).
+		For(&operatorv1.Operation{}).
+		Owns(&operatorv1.RuntimeTaskGroup{}).
+		Complete(r)
+
+	//TODO: watch DS for operation Daemonsets
+
+	r.recorder = mgr.GetEventRecorderFor("operation-controller")
+	return err
+}
+
 // Reconcile an operation
-func (r *OperationReconciler) Reconcile(req ctrl.Request) (ctrl.Result, error) {
-	_ = context.Background()
-	_ = r.Log.WithValues("operation", req.NamespacedName)
+func (r *OperationReconciler) Reconcile(req ctrl.Request) (_ ctrl.Result, rerr error) {
+	ctx := context.Background()
+	log := r.Log.WithValues("operation", req.NamespacedName)
+
+	// Fetch the Operation instance
+	operation := &operatorv1.Operation{}
+	if err := r.Client.Get(ctx, req.NamespacedName, operation); err != nil {
+		if apierrors.IsNotFound(err) {
+			return ctrl.Result{}, nil
+		}
+		return ctrl.Result{}, err
+	}
 
-	// your logic here
+	// Ignore the Operation if it is already completed or failed
+	if operation.Status.CompletionTime != nil {
```

Comment:
If it failed, will we get through here? (would `CompletionTime` be `!= nil`?) Because `reconcileDaemonSet` would then delete the daemonset. Asking because a failed daemon set could have valuable information for debugging.

Reply:
In case of an error we are not setting CompletionTime, so you have full access to the DaemonSet logs.